trieste/models/gpflow/sampler.py

```diff
@@ -348,84 +455,245 @@ def _prepare_theta_posterior_in_gram_space(self) -> tfp.distributions.Multivaria
+    aprroximate the Gaussian process posterior, a :class:`DecoupledTrajectorySampler` only
+    uses an RFF decomposition to approximate the Gausian process prior and instead using
+    a cannonical decomposition to discretize the effect of updating the prior on the given data.
 
-    def update_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunction:
-        """
-        Efficiently update a :const:`TrajectoryFunction` to reflect an update in its
-        underlying :class:`ProbabilisticModel` and resample accordingly.
+    In particular, we approximate the Gaussian processes' posterior samples as the finite feature
+    approximation
 
-        For a :class:`RandomFourierFeatureTrajectorySampler`, updating the sampler
-        corresponds to resampling the feature functions (taking into account any
-        changed kernel parameters) and recalculating the weight distribution.
+    .. math:: \hat{f}(.) = \sum_{i=1}^L w_i\phi_i(.) + \sum_{j=1}^m v_jk(.,z_j)
 
-        :param trajectory: The trajectory function to be resampled.
-        :return: The new resampled trajectory function.
+    where :math:`\phi_i(.)` and :math:`w_i` are the Fourier features and their weights that
+    discretize the prior. In contrast, `k(.,z_j)` and :math:`v_i` are the cannonical features and
+    their weights that discretize the data update.
+
+    The expression for :math:`v_i` depends on if we are using an exact Gaussian process or a sparse
+    approximations. See  eq. (13) in :cite:`wilson2020efficiently` for details.
+    """
+
+    def __init__(
+        self,
+        model: Union[
+            SupportsGetKernelGetInducingPoint,
+            SupportsGetKernelGetObservationNoiseSupportsGetInternalData,
+        ],
+        num_features: int = 1000,
+    ):
+        """
+        :param sample_size: The desired number of samples.
+        :param model: The model to sample from.
+        :param num_features: The number of features used to approximate the kernel. We use a default
+            of 1000 as it typically perfoms well for a wide range of kernels. Note that very smooth
+            kernels (e.g. RBF) can be well-approximated with fewer features.
+        :raise ValueError: If ``dataset`` is empty.
         """
-        tf.debugging.Assert(isinstance(trajectory, fourier_feature_trajectory), [])
 
-        if not hasattr(self._feature_functions, "_bias_init"):
-            # maintain support for gpflux 0.2.3 (but with retracing)
-            return self.get_trajectory()
-
-        bias_shape = tf.shape(self._feature_functions.b)
-        bias_dtype = self._feature_functions.b.dtype
-        weight_shape = tf.shape(self._feature_functions.W)
-        weight_dtype = self._feature_functions.W.dtype
-        self._feature_functions.b.assign(  # resample feature function's bias
-            self._feature_functions._bias_init(bias_shape, dtype=bias_dtype)
+        if not (
```

Comment:
isinstance supports tuples of types:

```suggestion
        if not isinstance(model, (SupportsGetKernelGetInducingPoint, SupportsGetKernelGetObservationNoiseSupportsGetInternalData)):
```